flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunDevBundleBuild.java

```
@@ -82,10 +89,154 @@ public static boolean needsBuild(File npmFolder) {
+            return true;
+        }
+
+        JsonObject packageJson = getPackageJson(npmFolder);
+        JsonObject statsJson = Json.parse(statsJsonContent);
+
+        // Get scanned @NpmPackage annotations
+        final Map<String, String> npmPackages = frontendDependencies
+                .getPackages();
+
+        if (!hashAndBundleModulesEqual(statsJson, packageJson, npmPackages)) {
+            // Hash in the project doesn't match the bundle hash or NpmPackages
+            // are found missing in bundle.
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Verify that package hash versions are equal and that all project
+     * npmPackages are in bundle.
+     *
+     *
+     * @param statsJson
+     *            devBundle statsJson
+     * @param packageJson
+     *            packageJson
+     * @param npmPackages
+     *            npm packages map
+     * @return {@code true} if up to date
+     */
+    private static boolean hashAndBundleModulesEqual(JsonObject statsJson,
+            JsonObject packageJson, Map<String, String> npmPackages) {
+
+        String packageJsonHash = getPackageJsonHash(packageJson);
+        String bundlePackageJsonHash = getStatsHash(statsJson);
+
+        if (packageJsonHash == null || packageJsonHash.isEmpty()) {
+            return false;
+        }
+
+        JsonObject bundleModules = statsJson.getObject("npmModules");
+
+        // Check that bundle modules contains all package dependencies
+        if (packageJsonHash.equals(bundlePackageJsonHash)) {
+            if (missingNpmPackages(npmPackages, bundleModules)) {
+                return false;
+            }
+        }
+
+        JsonObject dependencies = packageJson.getObject("dependencies");
+
+        List<String> collection = Arrays.stream(dependencies.keys())
+                .filter(pkg -> !dependencies.hasKey(pkg))
+                .collect(Collectors.toList());
```

This checks `dependencies` against `dependencies`, which gives an empty collection always. Should it check `bundleModules` agains `dependencies`?

---

fixed.

---

Do the tests need a tweak because of that ? :)

---

The test: `hashesMatch_statsMissingPackageJsonPackage_compilationRequired` tests that and the original mistake should have failed the test because the next part throws `java.lang.NullPointerException: Cannot invoke "elemental.json.JsonString.getString()" because the return value of "elemental.json.impl.JreJsonObject.get(String)" is null` Why the test passes even though something is thrown...

---

Oh. needsBuild eats exceptions and only logs them and says do rebuild

---

change test to use the internal method that doesn't catch